### DISCUSSION TITLE
Prisma unique key migrations and default values

### DISCUSSION THREAD

**Discussion by sklum** [planetscale/discussion#353] [Q&A] [answered]
Hi all, relatively new to `mysql` + `planetscale` + `prisma`, so apologies in advance if this is the wrong forum for these questions.  I have two:

1.  I'm currently trying to change the type and name of a unique id.  What’s the proper way to [change a single unique key](https://planetscale.com/docs/learn/change-single-unique-key) when using Prisma?  My assumption was, given an original schema:

```prisma
model Foo {
  oldId    Int    @id @default(autoincrement())
  name     String @db.VarChar(255)
}
```

we make a deploy request for:

```prisma
model Foo {
  newId    String @default(cuid())
  oldId    Int    @id @default(autoincrement())
  name     String @db.VarChar(255)
}
```

and then a deploy request for:

```prisma
model Foo {
  newId      String @default(cuid())
  name    String @db.VarChar(255)
}
```

Is this correct?

2.  More importantly, when I try to do this first deploy request on a table with production data, I get the following schema change:

```sql 
CREATE TABLE `Foo` (
+ `newId` varchar(191) NOT NULL,
  `oldId` int NOT NULL AUTO_INCREMENT,
  `name` varchar(255) NOT NULL,
  PRIMARY KEY (`oldId`)
 ) ENGINE InnoDB,
  CHARSET utf8mb4,
  COLLATE utf8mb4_unicode_ci;
```

It seems like the `@default` value for `newId` isn't being set to anything.  That aligns with the following error when trying to complete a deploy request (ignore the fact that `name` is empty for these entries, that's expected):

```
Field 'newId' doesn't have a default value (errno 1364) (sqlstate HY000) during query: insert into Foo(oldId,`name`) values (1,''), (2,''), (3,''), (4,''), (5,''), (6,''), (7,''), (8,''), (9,''), (10,''), (11,''), (12,''), (13,''), (14,''), (15,''), (16,''), (17,''), (18,''), (19,''), (20,''), (21,''), (22,''), (23,''), (24,''), (25,''), (26,''), (27,''), (28,''), (29,''), (30,''), (31,''), (32,''), (33,''), (34,''), (35,''), (36,''), (37,''), (38,''), (39,''), (40,''), (41,''), (42,''), (43,''), (44,''), (45,''), (46,''), (47,''), (48,''), (49,''), (50,''), (51,''), (52,''), (53,''), (54,''), (55,''), (56,''), (57,''), (58,''), (59,''), (60,''), (61,''), (62,''), (63,''), (64,''), (65,''), (66,''), (67,''), (68,''), (69,''), (70,''), (71,''), (72,''), (73,'')
```

What am I doing wrong here?  Is `@default(cuid())` not supported in some way?  I assume this is related to #144, but it's unclear to me how the value ended up being added in that case.

**Comment by janpio** [comment, marked as answer]
`@default(cuid())` is a Prisma level default that will be filled in by the Prisma Query Engine when you run a query via Prisma Client. It is not represented in the database.

**Reply by sklum** [reply]
Thanks.  So is there a way to do this migration from within the `prisma` schema or do I need to drop down to SQL commands?

**Reply by janpio** [reply]
Which migration do you mean exactly? The `cuid()` default is not supposed to be persisted in the database at this time. If you want to use a database level function you would need to use `@default(dbgenerated("..."))`. If you mean "generate valid CUID values for all existing rows", that would be indeed something to add to the migration file by hand (when using `prisma migrate dev`). Not sure how that works on PlanetScale.
